File: packages/shared-types/src/schemas.ts
import { z } from "zod";

export const studyModeSchema = z.enum([
  "standard",
  "adhd",
  "visual",
  "exam",
  "tutor",
  "speed-review",
]);

export type StudyMode = z.infer<typeof studyModeSchema>;

export const sourceSchema = z.object({
  id: z.string(),
  title: z.string(),
  shortTitle: z.string(),
  author: z.string().optional(),
  path: z.string(),
  type: z.enum(["textbook", "course-guide", "packet", "notes"]),
  pageCount: z.number().optional(),
  summary: z.string(),
  tags: z.array(z.string()),
  coverage: z.array(z.string()),
  uploadedAt: z.string(),
  lastProcessedAt: z.string().optional(),
});

export type Source = z.infer<typeof sourceSchema>;

export const topicReferenceSchema = z.object({
  id: z.string(),
  sourceId: z.string(),
  sourceTitle: z.string(),
  sourceShortTitle: z.string(),
  pageNumber: z.number(),
  heading: z.string(),
  excerpt: z.string(),
  highlightText: z.string().optional(),
});

export type TopicReference = z.infer<typeof topicReferenceSchema>;

export const sourcePageLineSchema = z.object({
  index: z.number(),
  text: z.string(),
  highlighted: z.boolean().default(false),
});

export const sourcePageSchema = z.object({
  source: sourceSchema,
  pageNumber: z.number(),
  totalPages: z.number(),
  heading: z.string().optional(),
  lines: z.array(sourcePageLineSchema),
  highlightText: z.string().optional(),
});

export type SourcePage = z.infer<typeof sourcePageSchema>;

export const sourceSelectionAssistResponseSchema = z.object({
  title: z.string(),
  response: z.string(),
  citations: z.array(topicReferenceSchema),
});

export type SourceSelectionAssistResponse = z.infer<typeof sourceSelectionAssistResponseSchema>;

export const formulaSchema = z.object({
  id: z.string(),
  label: z.string(),
  expression: z.string(),
  description: z.string(),
  variables: z.array(
    z.object({
      name: z.string(),
      meaning: z.string(),
    }),
  ),
  apTip: z.string(),
  topicIds: z.array(z.string()),
});

export type Formula = z.infer<typeof formulaSchema>;

export const graphPointSchema = z.object({
  id: z.string(),
  label: z.string(),
  x: z.number(),
  y: z.number(),
  color: z.string().optional(),
  description: z.string(),
  math: z.string(),
  intuition: z.string(),
});

export type GraphPoint = z.infer<typeof graphPointSchema>;

export const shadedAreaSchema = z.object({
  id: z.string(),
  label: z.string(),
  fill: z.string(),
  description: z.string(),
  math: z.string(),
  intuition: z.string(),
  pointIds: z.array(z.string()),
});

export type ShadedArea = z.infer<typeof shadedAreaSchema>;

export const graphCurveSchema = z.object({
  id: z.string(),
  label: z.string(),
  kind: z.enum(["line", "curve", "ppc", "step"]),
  color: z.string(),
  formula: z.string(),
  draggable: z.boolean().default(false),
});

export type GraphCurve = z.infer<typeof graphCurveSchema>;

export const graphModelSchema = z.object({
  id: z.string(),
  title: z.string(),
  type: z.enum([
    "supply-demand",
    "tax",
    "subsidy",
    "ppc",
    "monopoly",
    "cost-curves",
    "factor-market",
    "externality",
  ]),
  xAxisLabel: z.string(),
  yAxisLabel: z.string(),
  curves: z.array(graphCurveSchema),
  points: z.array(graphPointSchema),
  shadedAreas: z.array(shadedAreaSchema),
  formulas: z.array(z.string()),
  trapWarnings: z.array(z.string()),
});

export type GraphModel = z.infer<typeof graphModelSchema>;

export const flashcardSchema = z.object({
  id: z.string(),
  topicId: z.string(),
  front: z.string(),
  back: z.string(),
  difficulty: z.enum(["easy", "medium", "hard"]),
  tags: z.array(z.string()),
  dueAt: z.string().optional(),
  intervalDays: z.number().optional(),
  easeFactor: z.number().optional(),
  reviewCount: z.number().optional(),
  lastReviewedAt: z.string().optional(),
});

export type Flashcard = z.infer<typeof flashcardSchema>;

export const questionOriginSchema = z.enum(["official", "ap-like"]);

export const graphInteractionSchema = z.object({
  moduleId: z.string(),
  prompt: z.string(),
  targetType: z.enum(["point", "area"]),
  targetId: z.string(),
  targetLabel: z.string().optional(),
});

export type QuestionOrigin = z.infer<typeof questionOriginSchema>;
export type GraphInteraction = z.infer<typeof graphInteractionSchema>;

export const practiceQuestionSchema = z.object({
  id: z.string(),
  topicId: z.string(),
  unitId: z.string().optional(),
  type: z.enum(["mcq", "frq", "graph", "formula"]),
  origin: questionOriginSchema.default("ap-like"),
  stem: z.string(),
  prompt: z.string().optional(),
  choices: z.array(z.string()).optional(),
  answer: z.string(),
  explanation: z.string(),
  trap: z.string().optional(),
  difficulty: z.enum(["easy", "medium", "hard"]),
  sourceIds: z.array(z.string()),
  rubric: z.array(z.string()).default([]),
  graphInteraction: graphInteractionSchema.optional(),
});

export type PracticeQuestion = z.infer<typeof practiceQuestionSchema>;

export const noteSchema = z.object({
  id: z.string(),
  topicId: z.string(),
  title: z.string(),
  body: z.string(),
  pinnedGraphId: z.string().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export type Note = z.infer<typeof noteSchema>;

export const tutorMessageSchema = z.object({
  id: z.string(),
  role: z.enum(["user", "assistant", "system"]),
  content: z.string(),
  createdAt: z.string(),
});

export const tutorSessionSchema = z.object({
  id: z.string(),
  topicId: z.string().optional(),
  graphId: z.string().optional(),
  mode: studyModeSchema,
  messages: z.array(tutorMessageSchema),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export type TutorSession = z.infer<typeof tutorSessionSchema>;

export const progressTrackingSchema = z.object({
  id: z.string(),
  topicId: z.string(),
  confidence: z.number().min(0).max(100),
  mastery: z.number().min(0).max(100),
  lastStudiedAt: z.string().optional(),
  streakDays: z.number().default(0),
  weakSpots: z.array(z.string()),
  completedQuestionIds: z.array(z.string()),
});

export type ProgressTracking = z.infer<typeof progressTrackingSchema>;

export const practiceAttemptSchema = z.object({
  id: z.string(),
  questionId: z.string(),
  topicId: z.string(),
  mode: z.string(),
  answer: z.string(),
  correct: z.boolean(),
  confidence: z.number().min(0).max(100),
  durationSeconds: z.number().optional(),
  createdAt: z.string(),
});

export type PracticeAttempt = z.infer<typeof practiceAttemptSchema>;

export const reviewItemSchema = z.object({
  id: z.string(),
  itemType: z.string(),
  itemId: z.string(),
  topicId: z.string(),
  dueAt: z.string(),
  priority: z.number(),
  reason: z.string(),
  completed: z.boolean(),
});

export type ReviewItem = z.infer<typeof reviewItemSchema>;

export const dashboardOverviewSchema = z.object({
  focusTopicId: z.string().optional(),
  recommendedNextTopicId: z.string().optional(),
  sources: z.array(sourceSchema),
  progress: z.array(progressTrackingSchema),
  dueReviews: z.array(reviewItemSchema),
  weakTopicIds: z.array(z.string()),
});

export type DashboardOverview = z.infer<typeof dashboardOverviewSchema>;

export const ingestionStatsSchema = z.object({
  sourceCount: z.number(),
  topicCount: z.number(),
  chunkCount: z.number(),
  latestProcessedAt: z.string().optional(),
});

export type IngestionStats = z.infer<typeof ingestionStatsSchema>;

export const topicSchema = z.object({
  id: z.string(),
  slug: z.string(),
  unitId: z.string(),
  title: z.string(),
  shortDescription: z.string(),
  tenSecondMeaning: z.string(),
  whyItMatters: z.string(),
  apTrap: z.string(),
  graphIds: z.array(z.string()),
  formulaIds: z.array(z.string()),
  sourceIds: z.array(z.string()),
  keyIdeas: z.array(z.string()),
  examPhrasing: z.array(z.string()),
  visualAnchors: z.array(z.string()),
});

export type Topic = z.infer<typeof topicSchema>;

export const unitSchema = z.object({
  id: z.string(),
  number: z.number(),
  title: z.string(),
  examWeight: z.string(),
  description: z.string(),
  topicIds: z.array(z.string()),
});

export type Unit = z.infer<typeof unitSchema>;

export const searchResultSchema = z.object({
  id: z.string(),
  kind: z.enum(["topic", "formula", "question", "source", "graph"]),
  title: z.string(),
  summary: z.string(),
  href: z.string(),
  sourceIds: z.array(z.string()).default([]),
});

export type SearchResult = z.infer<typeof searchResultSchema>;

export const graphAnalyzerRequestSchema = z.object({
  imageUrl: z.string().optional(),
  uploadId: z.string().optional(),
  expectedGraphType: z.string().optional(),
});

export const graphAnalyzerResponseSchema = z.object({
  graphType: z.string(),
  confidence: z.number(),
  detectedAxes: z.array(z.string()),
  reconstructedGraphId: z.string().optional(),
  explanation: z.string(),
});

export type GraphAnalyzerRequest = z.infer<typeof graphAnalyzerRequestSchema>;
export type GraphAnalyzerResponse = z.infer<typeof graphAnalyzerResponseSchema>;

File: apps/api/app/models/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class SourceRecord(BaseModel):
    id: str
    title: str
    short_title: str
    path: str
    type: Literal["textbook", "course-guide", "packet", "notes"]
    summary: str
    tags: list[str] = Field(default_factory=list)
    page_count: int | None = None
    uploaded_at: datetime = Field(default_factory=datetime.utcnow)
    last_processed_at: datetime | None = None


class SectionBlock(BaseModel):
    heading: str
    page: int
    text: str


class DocumentChunk(BaseModel):
    id: str
    source_id: str
    page_start: int
    page_end: int
    heading: str
    text: str
    estimated_tokens: int
    topic_candidates: list[str] = Field(default_factory=list)
    graph_mentions: list[str] = Field(default_factory=list)
    formula_mentions: list[str] = Field(default_factory=list)


class TopicBundle(BaseModel):
    id: str
    unit_id: str | None = None
    topic_slug: str
    title: str
    summary: str
    source_ids: list[str] = Field(default_factory=list)
    chunk_ids: list[str] = Field(default_factory=list)
    explanation_variants: list[str] = Field(default_factory=list)
    graph_mentions: list[str] = Field(default_factory=list)
    formula_mentions: list[str] = Field(default_factory=list)


class UploadResponse(BaseModel):
    message: str
    source_ids: list[str]
    processed_count: int


class SearchHit(BaseModel):
    id: str
    kind: Literal["topic", "chunk", "source"]
    title: str
    summary: str
    score: float
    source_ids: list[str] = Field(default_factory=list)


class PracticeQuestionRecord(BaseModel):
    id: str
    topic_id: str
    unit_id: str | None = None
    type: Literal["mcq", "frq", "graph", "formula"]
    origin: Literal["official", "ap-like"] = "ap-like"
    stem: str
    prompt: str | None = None
    choices: list[str] | None = None
    answer: str
    explanation: str
    trap: str | None = None
    difficulty: Literal["easy", "medium", "hard"]
    source_ids: list[str] = Field(default_factory=list)
    rubric: list[str] = Field(default_factory=list)
    graph_interaction: dict | None = None


class FlashcardRecord(BaseModel):
    id: str
    topic_id: str
    front: str
    back: str
    difficulty: Literal["easy", "medium", "hard"]
    tags: list[str] = Field(default_factory=list)
    due_at: datetime | None = None
    interval_days: int = 1
    ease_factor: float = 2.5
    review_count: int = 0
    last_reviewed_at: datetime | None = None


class NoteRecord(BaseModel):
    id: str
    topic_id: str
    title: str
    body: str
    pinned_graph_id: str | None = None
    bookmarked: bool = False
    created_at: datetime
    updated_at: datetime


class NoteUpsert(BaseModel):
    title: str
    body: str
    pinned_graph_id: str | None = None
    bookmarked: bool = False


class TutorMessageRecord(BaseModel):
    id: str
    role: Literal["user", "assistant", "system"]
    content: str
    graph_context: dict | None = None
    created_at: datetime


class TutorSessionRecord(BaseModel):
    id: str
    topic_id: str | None = None
    graph_id: str | None = None
    mode: str
    graph_state: dict | None = None
    created_at: datetime
    updated_at: datetime
    messages: list[TutorMessageRecord] = Field(default_factory=list)


class TutorSessionCreate(BaseModel):
    topic_id: str | None = None
    graph_id: str | None = None
    mode: str = "tutor"
    graph_state: dict | None = None


class TutorMessageCreate(BaseModel):
    content: str
    role: Literal["user"] = "user"
    graph_context: dict | None = None


class ProgressRecord(BaseModel):
    id: str
    topic_id: str
    confidence: int
    mastery: int
    last_studied_at: datetime | None = None
    streak_days: int = 0
    weak_spots: list[str] = Field(default_factory=list)
    completed_question_ids: list[str] = Field(default_factory=list)


class PracticeAttemptCreate(BaseModel):
    question_id: str
    topic_id: str
    mode: str = "standard"
    answer: str
    confidence: int = 50
    duration_seconds: int | None = None


class PracticeAttemptRecord(BaseModel):
    id: str
    question_id: str
    topic_id: str
    mode: str
    answer: str
    correct: bool
    confidence: int
    duration_seconds: int | None = None
    created_at: datetime


class FlashcardReviewCreate(BaseModel):
    flashcard_id: str
    confidence: int = 70


class FlashcardImportRequest(BaseModel):
    topic_id: str | None = None
    raw_text: str
    source: Literal["quizlet", "manual"] = "quizlet"
    difficulty: Literal["easy", "medium", "hard"] = "medium"
    tags: list[str] = Field(default_factory=list)


class FlashcardImportResponse(BaseModel):
    imported_count: int
    flashcards: list[FlashcardRecord] = Field(default_factory=list)


class ReviewItemRecord(BaseModel):
    id: str
    item_type: str
    item_id: str
    topic_id: str
    due_at: datetime
    priority: int
    reason: str
    completed: bool = False


class DashboardOverview(BaseModel):
    focus_topic_id: str | None = None
    recommended_next_topic_id: str | None = None
    sources: list[SourceRecord] = Field(default_factory=list)
    progress: list[ProgressRecord] = Field(default_factory=list)
    due_reviews: list[ReviewItemRecord] = Field(default_factory=list)
    weak_topic_ids: list[str] = Field(default_factory=list)


class IngestionStats(BaseModel):
    source_count: int
    topic_count: int
    chunk_count: int
    latest_processed_at: datetime | None = None


class TopicReferenceRecord(BaseModel):
    id: str
    source_id: str
    source_title: str
    source_short_title: str
    page_number: int
    heading: str
    excerpt: str
    highlight_text: str | None = None


class SourcePageLine(BaseModel):
    index: int
    text: str
    highlighted: bool = False


class SourcePageRecord(BaseModel):
    source: SourceRecord
    page_number: int
    total_pages: int
    heading: str | None = None
    lines: list[SourcePageLine] = Field(default_factory=list)
    highlight_text: str | None = None


class SourceSelectionAssistRequest(BaseModel):
    text: str
    action: Literal["summarize", "explain-simple", "link-to-apmicro", "tutor-help"]
    source_id: str | None = None
    page_number: int | None = None
    topic_slug: str | None = None


class SourceSelectionAssistResponse(BaseModel):
    title: str
    response: str
    citations: list[TopicReferenceRecord] = Field(default_factory=list)


class GraphAnalyzerRequest(BaseModel):
    upload_id: str | None = None
    image_url: str | None = None
    expected_graph_type: str | None = None


class GraphAnalyzerResponse(BaseModel):
    graph_type: str
    confidence: float
    detected_axes: list[str]
    reconstructed_graph_id: str | None = None
    explanation: str
